src/encoded/static/components/facets.js

```diff
@@ -0,0 +1,518 @@
+    }
+
+    _addGenomeFilter() {
+        this.props.addGenomeFilter(this.props.buttonLabel, this.props.facetLabel);
+    }
+
+    render() {
+        return (
+            <button
+                className="browser-filter"
+                onClick={this._addGenomeFilter}
+                key={this.props.facetKey}
+            >
+                <i className="icon icon-times-circle" />
+                {this.props.buttonLabel}
+            </button>
+        );
+    }
+}
+
+FilterButton.propTypes = {
+    buttonLabel: PropTypes.string.isRequired,
+    facetLabel: PropTypes.string.isRequired,
+    facetKey: PropTypes.string.isRequired,
+    addGenomeFilter: PropTypes.func.isRequired,
+};
+
+
+// The "FacetList" component is used on the genome browser view
+
+
+export class FacetList extends React.Component {
+    constructor() {
+        super();
+
+        this.state = {
+            selectedFacets: [],
+            unsanitizedSearchTerms: {},
+            searchTerms: {},
+        };
+        this.addGenomeFilter = this.addGenomeFilter.bind(this);
+        this.toggleFacetDisplay = this.toggleFacetDisplay.bind(this);
+        this.clearGenomeFilters = this.clearGenomeFilters.bind(this);
+        this.handleSearch = this.handleSearch.bind(this);
+    }
+
+    componentDidMount() {
+        this.setState({ selectedFacets: this.props.selectedFilters });
+    }
+
+    addGenomeFilter(item, category) {
+        const itemFacet = `${item}AND${category}`;
+        if (this.state.selectedFacets.includes(itemFacet)) {
+            this.setState(prevState => ({
+                selectedFacets: prevState.selectedFacets.filter(facet => facet !== itemFacet),
+            }), () => {
+                this.props.handleFacetList(this.state.selectedFacets);
+            });
+        } else {
+            this.setState(prevState => ({
+                selectedFacets: [...prevState.selectedFacets, itemFacet],
+            }), () => {
+                this.props.handleFacetList(this.state.selectedFacets);
+            });
+        }
+    }
+
+    clearGenomeFilters() {
+        this.setState({ selectedFacets: [] }, () => {
+            this.props.handleFacetList(this.state.selectedFacets);
+        });
+    }
+
+    toggleFacetDisplay() {
+        this.setState(prevState => ({
+            facetDisplay: !(prevState.facetDisplay),
+        }));
+    }
+
+    handleSearch(e, typeaheadIdentifier) {
+        const filterVal = String(globals.sanitizedString(e.target.value));
+        const targetValue = e.target.value;
+        this.setState((prevState) => {
+            const unsanitizedSearchTerms = { ...prevState.unsanitizedSearchTerms };
+            const searchTerms = { ...prevState.searchTerms };
+            unsanitizedSearchTerms[typeaheadIdentifier] = targetValue;
+            searchTerms[typeaheadIdentifier] = filterVal;
+            return {
+                searchTerms,
+                unsanitizedSearchTerms,
+            };
+        });
+    }
```

Comment:
I don’t think you need both `unsanitizedSearchTerms` _and_ `searchTerms` to be in state. Just calculate `searchTerms` when you need it, in `render()` — basically the philosophy Dan Abramov talks about here, in which there should be a high bar before putting anything in state: https://twitter.com/dan_abramov/status/749710501916139520?s=20&t=7-uNRnEgre6Im48YM-GtyQ